### PR TITLE
ceph: check for ceph object user cleanup

### LINKED ISSUE #5712: ceph: object store deletion should blocked until all the ceph object store users is deleted
**Is this a bug report or feature request?**
* Bug Report

Currently objectstore waits for all the obcs need to be deleted, but does not wait for object store users, need to add that check
**How to reproduce it (minimal and precise):**

**File(s) to submit**:

* Cluster CR (custom resource), typically called `cluster.yaml`, if necessary
* Operator's logs, if necessary
* Crashing pod(s) logs, if necessary

 To get logs, use `kubectl -n <namespace> logs <pod name>`
When pasting logs, always surround them with backticks or use the `insert code` button from the Github UI.
Read [Github documentation if you need help](https://help.github.com/en/articles/creating-and-highlighting-code-blocks).

**Environment**:
* OS (e.g. from /etc/os-release):
* Kernel (e.g. `uname -a`):
* Cloud provider or hardware configuration:
* Rook version (use `rook version` inside of a Rook Pod):
* Storage backend version (e.g. for ceph do `ceph -v`):
* Kubernetes version (use `kubectl version`):
* Kubernetes cluster type (e.g. Tectonic, GKE, OpenShift):
* Storage backend status (e.g. for Ceph use `ceph health` in the [Rook Ceph toolbox](https://rook.io/docs/rook/master/ceph-toolbox.html)):

### PR DESCRIPTION
**Pull request by thotz**
**Description of your changes:**
Do not delete ceph object store until ceph object users are cleaned up

**Which issue is resolved by this Pull Request:**
Resolves #5712

**Checklist:**

- [ ] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/master/development-flow.html#commit-structure).
- [ ] **Skip Tests for Docs**: Add the flag for skipping the build if this is only a documentation change. See [here](https://github.com/rook/rook/blob/master/INSTALL.md#skip-ci) for the flag.
- [ ] **Skip Unrelated Tests**: Add a flag to run tests for a specific storage provider. See [test options](https://github.com/rook/rook/blob/master/INSTALL.md#test-storage-provider).
- [ ] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/master/development-flow.html#submitting-a-pull-request)
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.
- [ ] Pending release notes updated with breaking and/or notable changes, if necessary.
- [ ] Upgrade from previous release is tested and upgrade user guide is updated, if necessary.
- [ ] Code generation (`make codegen`) has been run to update object specifications, if necessary.

### CODE REVIEW

File: pkg/operator/ceph/object/controller.go [hunk 1]
```diff
@@ -483,3 +490,24 @@ func (r *ReconcileCephObjectStore) startMonitoring(objectstore *cephv1.CephObjec
 	logger.Info("starting rgw healthcheck")
 	go rgwChecker.checkObjectStore(r.objectStoreChannels[objectstore.Name].stopChan)
 }
+
+func (r *ReconcileCephObjectStore) verifyObjectUserCleanup(objectstore *cephv1.CephObjectStore) (reconcile.Result, bool) {
```
Comment (travisn) [inline pkg/operator/ceph/object/controller.go:494]: please add a unit test for this method. It should be fairly straight forward to create the mocked object users and confirm if the return is true or false based on whether there are users.

Comment (thotz) [reply pkg/operator/ceph/object/controller.go:494]: Done

File: pkg/operator/ceph/object/controller_test.go [hunk 2]
```diff
@@ -330,4 +330,28 @@ func TestCephObjectStoreController(t *testing.T) {
 	assert.NoError(t, err)
 	assert.Equal(t, cephv1.ConditionReady, objectStore.Status.Phase, objectStore)
 	logger.Info("PHASE 3 DONE")
+
+	// Test the functionality of verifyObjectUserCleanup
+	// function returns true if there are no cephobjectusers,
+	// otherwise false
+	logger.Info("STARTING PHASE 4")
+	_, userexists := r.verifyObjectUserCleanup(objectStore)
```
Comment (travisn) [inline pkg/operator/ceph/object/controller_test.go:338]: The var `userexists` seems backwards to me. What about this?
```suggestion
	_, okToDelete := r.verifyObjectUserCleanup(objectStore)
```

File: pkg/operator/ceph/object/controller_test.go [hunk 3]
```diff
@@ -330,4 +330,28 @@ func TestCephObjectStoreController(t *testing.T) {
...
+			Store: store,
+		},
+		TypeMeta: metav1.TypeMeta{
+			Kind: "CephObjectStoreUser",
+		},
+	}
+	objectUser, _ = r.context.RookClientset.CephV1().CephObjectStoreUsers(objectStore.Namespace).Create(objectUser)
+	_, userexists = r.verifyObjectUserCleanup(objectStore)
+	assert.False(t, userexists)
+	logger.Info("PHASE 4 DONE")
```
Comment (travisn) [inline pkg/operator/ceph/object/controller_test.go:356]: nit: The log messages should indicate what happened, rather than a phase step number.

File: pkg/operator/ceph/object/controller.go [hunk 4]
```diff
@@ -246,6 +246,13 @@ func (r *ReconcileCephObjectStore) reconcile(request reconcile.Request) (reconci
 			return response, nil
 		}
 
+		response, ok = r.verifyObjectUserCleanup(cephObjectStore)
```
Comment (travisn) [inline pkg/operator/ceph/object/controller.go:249]:
```suggestion
		response, okToDelete = r.verifyObjectUserCleanup(cephObjectStore)
```

File: pkg/operator/ceph/object/controller_test.go [hunk 5]
```diff
@@ -330,4 +330,28 @@ func TestCephObjectStoreController(t *testing.T) {
 	assert.NoError(t, err)
 	assert.Equal(t, cephv1.ConditionReady, objectStore.Status.Phase, objectStore)
 	logger.Info("PHASE 3 DONE")
+
+	// Test the functionality of verifyObjectUserCleanup
+	// function returns true if there are no cephobjectusers,
+	// otherwise false
```
Comment (leseb) [inline pkg/operator/ceph/object/controller_test.go:336]: We need 2 tests, one, when there is/, are CephObjectStoreUsers and one when there is none.

File: pkg/operator/ceph/object/controller.go [hunk 6]
```diff
@@ -483,3 +490,24 @@ func (r *ReconcileCephObjectStore) startMonitoring(objectstore *cephv1.CephObjec
 	logger.Info("starting rgw healthcheck")
 	go rgwChecker.checkObjectStore(r.objectStoreChannels[objectstore.Name].stopChan)
 }
+
+func (r *ReconcileCephObjectStore) verifyObjectUserCleanup(objectstore *cephv1.CephObjectStore) (reconcile.Result, bool) {
+	cephObjectUsers, err := r.context.RookClientset.CephV1().CephObjectStoreUsers(objectstore.Namespace).List(metav1.ListOptions{})
```
Comment (leseb) [inline pkg/operator/ceph/object/controller.go:495]: Nit, feel free to take it or not but with controller runtime, I believe it's a good habit to start using `client.Client` instead of the `RookClientset`, feature-wise it's the same though.

Comment (thotz) [reply pkg/operator/ceph/object/controller.go:495]: @leseb Okay. But I was not able to `CephObjectStoreUsers(objectstore.Namespace).List(metav1.ListOptions{})` api from `r.client.`
Am I looking in the wrong places?

File: pkg/operator/ceph/object/controller_test.go (hunk 5 above)
Comment (thotz) [reply pkg/operator/ceph/object/controller_test.go:336]: Done

File: pkg/operator/ceph/object/controller_test.go (hunk 3 above)
Comment (thotz) [reply pkg/operator/ceph/object/controller_test.go:356]: Done
